Document server side field validation (#30532)
kubernetes/website: content/en/docs/reference/kubernetes-api/common-parameters/common-parameters.md

> 

Reviewer: huh... the header says this is auto-generated:
`The file is auto-generated from the Go source code of the component using a generic...`

not sure if that's true or where the source for this is

Reviewer: current content lgtm though

Author: hmm, that's a good question. The other parameter descriptions do look like they are pulled from the comments of `meta/v1/types.go`. 

I tried following the instructions for [updating the generated docs](https://github.com/kubernetes/website/blob/main/content/en/docs/contribute/generate-ref-docs/quickstart.md) and was successfully able to run `./update-imported-docs.py`, but all it did was update `kubectl-commands.html` with an unrelated change. I think one issue may be that I can't use a branch that [doesn't exist yet ](https://github.com/kubernetes/website/blob/main/update-imported-docs/reference.yml#L10)

@mehabhalodiya can you confirm that these docs are still being auto-generated and that the source of truth is `meta/v1/types.go`? If so, do you know how to best update these so that my [new changes](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L557-L568) are included here?

Reviewer 2: @kevindelgado I too believe, `The file is auto-generated` as mentioned by @liggitt @sftim .
Also, may be you can set the branch of the previous release until 1.23 isn't formally released.

## fieldValidation {#fieldValidation}

> 

Reviewer: this should indicate the feature state and release:

```
{{< feature-state for_k8s_version="1.23" state="alpha" >}}
```

## fieldValidation {#fieldValidation}

> {{< feature-state for_k8s_version="v1.23" state="alpha" >}}

Reviewer 3: Should we mention that we're only ever returning strict errors when there is no non-strict errors?

When both the fieldValidation parameter is present and the `ServerSideFieldValidation` feature gate is enabled, it instructs the server on how to handle objects in the request (POST/PUT/PATCH) containing unknown or duplicate fields. Valid values are:
- Ignore: This will ignore any unknown fields that are silently dropped from the
  object, and will ignore all but the last duplicate field that the decoder
  encounters. This is the default behavior prior to v1.23 and is the default
  behavior when the `ServerSideFieldValidation` feature gate is disabled.
- Warn: This will send a warning via the [standard warning response
  header](https://datatracker.ietf.org/doc/html/rfc7234#section-5.5) for each

>   unknown that is dropped from the object, and for each duplicate field encountered.

Reviewer 3: each unknown *fields*, I'm assuming

  The request will still succeed if there are no other errors, and will only persist
  the last of any duplicate fields. This is the default
  when the `ServerSideFieldValidation` feature gate is enabled.
- Strict: This will fail the request with a BadRequest error if any unknown
  fields would be dropped from the object, or if any duplicate fields are
  present. The error returned from the server will contain
  all unknown and duplicate fields encountered.

<hr>

> 

Reviewer 3: That's a lot of empty lines :-)

## fieldValidation {#fieldValidation}

> When both the fieldValidation parameter is present and the `ServerSideFieldValidation` feature gate is enabeld, it instructs the server on how to handle objects in the request (POST/PUT/PATCH) containing unknown or duplicate fields. Valid values are:

Reviewer: ```suggestion
When both the fieldValidation parameter is present and the `ServerSideFieldValidation` feature gate is enabled, it instructs the server on how to handle objects in the request (POST/PUT/PATCH) containing unknown or duplicate fields. Valid values are:
```

> - Ignore: This will ignore any unknown fields, silently dropping them from the

Reviewer: ```suggestion
- Ignore: This will ignore any unknown fields that are silently dropped from the
```

  object, and ignoring all but the last duplicate field that the decoder
  encounters. This is the default behavior prior to v1.23 and is the default
  behavior when the `ServerSideFieldValidation` feature gate is disabled.
- Warn: This will send a warning via the [standarding warning response
  header](https://datatracker.ietf.org/doc/html/rfc7234#section-5.5) for each

>   unknown or duplicate field it encounters. The request will still succeed if

Reviewer: ```suggestion
  unknown field that is dropped from the object, and for each duplicate field encountered. The request will still succeed if
```

>   there are no other errors, the resulting object will drop any unknown
>   fields and only persist the last of any duplicate fields. This is the default

Reviewer: ```suggestion
  there are no other errors, and only persist the last of any duplicate fields. This is the default
```

  when the `ServerSideFieldValidation` feature gate is enabled.

> - Strict: This will fail the request with a BadRequest error if any unknown or
>   duplicate fields are present. The error returned from the server will contain

Reviewer: ```suggestion
- Strict: This will fail the request with a BadRequest error if any unknown fields
  would be dropped from the object, or if any duplicate fields are present. The error returned from the server will contain
```

  all unknown and duplicate fields encountered.

One exception is that CRDs with the `x-kubernetes-preserve-unknown-fields`
property will continue to preserve the unknown fields and will not indicate via
warning/error if an unknown field is detected for any node in the custom

> resource that has `x-kubernetes-preserve-unknown-fields` set.

Reviewer: this is probably confusing to mention here... I would frame the "unknown fields" bits above as "unknown fields that would be [silently] dropped from the object" and omit this
